safe_transaction_service/history/models.py

```diff
@@ -1606,6 +1606,70 @@ class SafeContractManager(models.Manager):
 
+    def get_count_relevant_txs_for_safe(self, address: ChecksumAddress) -> int:
+        """
+        This method searches multiple tables and count every tx or event for a Safe.
+        It will return the same or higher value if compared to counting ``get_all_tx_identifiers``
+        as that method will group some transactions (for example, 3 ERC20 can be grouped in a ``MultisigTransaction``,
+        so it will be ``1`` element for ``get_all_tx_identifiers`` but ``4`` for this function.
+
+        This query should be pretty fast, and it's meant to be used for invalidating caches.
+
+        :param address:
+        :return: number of relevant txs for a Safe
+        """
+
+        query = """
+                SELECT COUNT(*)
+                FROM (
+                    -- Get multisig transactions
+                    SELECT 1
+                    FROM "history_multisigtransaction"
+                    WHERE "history_multisigtransaction"."safe" = %s
+                    UNION ALL
+                    -- Get confirmations
+                    SELECT 1
+                    FROM "history_multisigtransaction"
+                       JOIN "history_multisigconfirmation" ON "history_multisigtransaction"."safe_tx_hash" = "history_multisigconfirmation"."multisig_transaction_id"
+                    WHERE "history_multisigtransaction"."safe" = %s
+                    UNION ALL
+                    -- Get ERC20 Transfers
+                    SELECT 1
+                    FROM "history_erc20transfer"
+                    WHERE (
+                            "history_erc20transfer"."to" = %s
+                            OR "history_erc20transfer"."_from" = %s
+                        )
+                    UNION ALL
+                    -- Get ERC721 Transfers
+                    SELECT 1
+                    FROM "history_erc721transfer"
+                    WHERE (
+                            "history_erc721transfer"."to" = %s
+                            OR "history_erc721transfer"."_from" = %s
+                        )
+                    UNION ALL
+                    -- Get Ether Transfers
+                    SELECT 1
+                    FROM "history_internaltx"
+                    WHERE (
+                            "history_internaltx"."call_type" = 0
+                            AND "history_internaltx"."to" = %s
+                            AND "history_internaltx"."value" > 0
+                        )
+                    UNION ALL
+                    -- Get Module Transactions
+                    SELECT 1
+                    FROM "history_moduletransaction"
+                    WHERE "history_moduletransaction"."safe" = %s
+                ) subquery
+                """
+
+        with connection.cursor() as cursor:
+            hex_address = HexBytes(address)
+            cursor.execute(query, [hex_address] * 8)
+            return cursor.fetchone()[0]
+
```

Comment:
Can you analyze this query with a safe with a lot of internalTx, multisigTx...?

Reply:
Done, with the new index less than one second for the biggest Safes in our database